brillig_vm/src/lib.rs

```diff
@@ -923,6 +943,71 @@ mod tests {
         assert_eq!(vm.foreign_call_counter, 1);
     }
 
+    /// Calling a simple foreign call function that takes any string input and concatenates it with itself
+    #[test]
+    fn foreign_call_opcode_vector_input_and_output() {
+        let r_input_pointer = RegisterIndex::from(0);
+        let r_input_size = RegisterIndex::from(1);
+        // We need to pass a location of appropriate size
+        let r_output_pointer = RegisterIndex::from(2);
+        let r_output_size = RegisterIndex::from(3);
+
+        // Our first string to use the identity function with
+        let input_string =
+            vec![Value::from(1u128), Value::from(2u128), Value::from(3u128), Value::from(4u128)];
+        // Double the string (concatenate it with itself)
+        let output_string: Vec<Value> =
+            input_string.iter().cloned().chain(input_string.clone()).collect();
+
+        // First call:
+        let string_double_program = vec![
+            // input_pointer = 0
+            Opcode::Const { destination: r_input_pointer, value: Value::from(0u128) },
+            // input_size = input_string.len() (constant here)
+            Opcode::Const { destination: r_input_size, value: Value::from(input_string.len()) },
+            // output_pointer = 0 + input_size = input_size
+            Opcode::Const { destination: r_output_pointer, value: Value::from(input_string.len()) },
+            // output_size = input_size * 2
+            Opcode::Const {
+                destination: r_output_size,
+                value: Value::from(input_string.len() * 2),
+            },
+            // output_pointer[0..output_size] = string_double(input_pointer[0...input_size])
+            Opcode::ForeignCall {
+                function: "string_double".into(),
+                destinations: vec![RegisterOrMemory::HeapVector(r_output_pointer, r_output_size)],
+                inputs: vec![RegisterOrMemory::HeapVector(r_input_pointer, r_input_size)],
+            },
+        ];
+
+        let mut vm = brillig_execute_and_get_vm(input_string.clone(), string_double_program);
+
+        // Check that VM is waiting
+        assert_eq!(
+            vm.status,
+            VMStatus::ForeignCallWait {
+                function: "string_double".into(),
+                inputs: vec![input_string]
+            }
+        );
+
+        // Push result we're waiting for
+        vm.foreign_call_results.push(ForeignCallResult { values: vec![output_string.clone()] });
+
+        // Resume VM
+        brillig_execute(&mut vm);
+
+        // Check that VM finished once resumed
+        assert_eq!(vm.status, VMStatus::Finished);
+
+        // Check result in memory
+        let result_values = vm.memory[0..output_string.len()].to_vec();
```

Comment:
We cloned the input string twice exactly so 0..output_string.len() would trivially match the output string. If we reversed the output_string after constructing it (before resolving the foreign call) we would get a failure as this is reading the input string and first half of the output string. Switching this to `input_string.len()..(input_string.len() + output_string.len())` makes the test more accurate

Reply:
I ended up pushing a commit directly here that changed this test to reverse the concatenation (https://github.com/noir-lang/acvm/pull/370/commits/dd05610dcdab4f1f383e42f772c829ab22f2293d) to make the test more thorough